src/Razor/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/CodeWriter.cs

```diff
@@ -62,16 +72,43 @@ public string NewLine
             }
         }
 
-        // Internal for testing.
-        internal CodeWriter Indent(int size)
+        public CodeWriter Indent(int size)
         {
-            if (Length == 0 || this[Length - 1] == '\n')
+            if (size == 0 || (Length != 0 && this[Length - 1] != '\n'))
             {
-                _builder.Append(' ', size);
+                return this;
+            }
 
-                _currentLineCharacterIndex += size;
-                _absoluteIndex += size;
+            var actualSize = 0;
+            if (IndentWithTabs)
+            {
+                // Avoid writing directly to the StringBuilder here, that will throw off the manual indexing 
+                // done by the base class.
+                var tabs = size / TabSize;
+                actualSize += tabs;
+                for (var i = 0; i < tabs; i++)
+                {
+                    _builder.Append("\t");
+                }
```

Comment:
Could be simplified to:
```suggestion
                _builder.Append('\t', tabs);
```
(also in the two places below)

Reply:
Thanks! Updated.